improver/threshold.py

```diff
@@ -284,95 +544,165 @@ def process(self, input_cube: Cube) -> Cube:
-            # otherwise, scale exceedance probabilities linearly between 0/1
-            # at the min/max fuzzy bounds and 0.5 at the threshold value
-            else:
-                truth_value = np.where(
-                    cube.data < threshold,
-                    rescale(
-                        cube.data,
-                        data_range=(bounds[0], threshold),
-                        scale_range=(0.0, 0.5),
-                        clip=True,
-                    ),
-                    rescale(
-                        cube.data,
-                        data_range=(threshold, bounds[1]),
-                        scale_range=(0.5, 1.0),
-                        clip=True,
-                    ),
-                )
-                # if requirement is for probabilities less_than or
-                # less_than_or_equal_to the threshold (rather than
-                # greater_than or greater_than_or_equal_to), invert
-                # the exceedance probability
-                if "less_than" in self.comparison_operator.spp_string:
-                    truth_value = 1.0 - truth_value
+        # Slice over realizations if required and create an empty threshold
+        # cube to store the resulting thresholded data.
+        if self.collapse_coord is not None:
+            input_slices = input_cube.slices_over(self.collapse_coord)
+            thresholded_cube = self._create_threshold_cube(
+                next(input_cube.slices_over(self.collapse_coord))
+            )
+        else:
+            input_slices = [input_cube]
+            thresholded_cube = self._create_threshold_cube(input_slices[0])
+
+        # Create a zeroed array for storing contributions (i.e. number of
+        # unmasked realization values contributing to calculation).
+        contribution_total = np.zeros(
+            next(thresholded_cube.slices_over(self.threshold_coord_name)).shape,
+            dtype=int,
+        )
 
-            truth_value = truth_value.astype(FLOAT_DTYPE)
+        for cube in input_slices:
+            # Tests performed on each slice rather than whole cube to avoid
+            # realising all of the data.
+            if np.isnan(cube.data).any():
+                raise ValueError("Error: NaN detected in input cube data")
 
             if np.ma.is_masked(cube.data):
-                # update unmasked points only
-                cube.data[~input_cube.data.mask] = truth_value[~input_cube.data.mask]
+                mask = cube.data.mask
+                unmasked = ~mask
             else:
-                cube.data = truth_value
-
-            self._add_threshold_coord(cube, threshold)
-            cube.coord(var_name="threshold").convert_units(input_cube.units)
-            self._update_metadata(cube)
+                unmasked = np.ones(cube.shape, dtype=bool)
+
+            fill_value = netCDF4.default_fillvals.get(cube.dtype.str[1:], np.inf)
```

Comment:
I guess because it's the fastest method and there is no danger of realizing the data. It takes 1 micro-second as it just extracts a value from a dictionary. Even the `np.ma.default_fill_value(array.dtype)` method is slower.

I've moved this into the `maximum_within_vicinity` method so it is no longer passed around. As it's no longer an argument I think your other concerns are moot.